server/libs/infra/src/search/typesense.repository.ts

```diff
@@ -0,0 +1,322 @@
+      });
+      if (_items.length > 0) {
+        await this.client
+          .collections(schema.name)
+          .documents()
+          .import(_items, { action: 'upsert', dirty_values: 'coerce_or_drop' });
+      }
+      if (done) {
+        await this.updateAlias(collection);
+      }
+    } catch (error: any) {
+      this.handleError(error);
+    }
+  }
+
+  search(collection: SearchCollection.ASSETS, query: string, filter: SearchFilter): Promise<SearchResult<AssetEntity>>;
+  search(collection: SearchCollection.ALBUMS, query: string, filter: SearchFilter): Promise<SearchResult<AlbumEntity>>;
+  async search(collection: SearchCollection, query: string, filters: SearchFilter) {
+    const alias = await this.client.aliases(collection).retrieve();
+
+    const { userId } = filters;
+
+    const _filters = [`ownerId:${userId}`];
+
+    if (collection === SearchCollection.ASSETS) {
+      for (const item of schemaMap[collection].fields || []) {
+        let value = filters[item.name as keyof SearchFilter];
+        if (Array.isArray(value)) {
+          value = `[${value.join(',')}]`;
+        }
+        if (item.facet && value !== undefined) {
+          _filters.push(`${item.name}:${value}`);
+        }
+      }
+
+      this.logger.debug(`Searching query='${query}', filters='${JSON.stringify(_filters)}'`);
+
+      const results = await this.client
+        .collections<AssetEntity>(alias.collection_name)
+        .documents()
+        .search({
+          q: query,
+          query_by: [
+            'exifInfo.imageName',
+            'exifInfo.country',
+            'exifInfo.state',
+            'exifInfo.city',
+            'exifInfo.description',
+            'smartInfo.tags',
+            'smartInfo.objects',
+          ].join(','),
+          filter_by: _filters.join(' && '),
+          per_page: 250,
+          facet_by: (assetSchema.fields || [])
+            .filter((field) => field.facet)
+            .map((field) => field.name)
+            .join(','),
+        });
+
+      return this.asResponse(results);
+    }
+
+    if (collection === SearchCollection.ALBUMS) {
+      const results = await this.client
+        .collections<AlbumEntity>(alias.collection_name)
+        .documents()
+        .search({
+          q: query,
+          query_by: 'albumName',
+          filter_by: _filters.join(','),
+        });
+
+      return this.asResponse(results);
+    }
+
+    throw new Error(`Invalid collection: ${collection}`);
+  }
+
+  private asResponse<T extends DocumentSchema>(results: SearchResponse<T>): SearchResult<T> {
+    return {
+      page: results.page,
+      total: results.found,
+      count: results.out_of,
+      items: (results.hits || []).map((hit) => hit.document),
+      facets: (results.facet_counts || []).map((facet) => ({
+        counts: facet.counts.map((item) => ({ count: item.count, value: item.value })),
+        fieldName: facet.field_name as string,
+      })),
+    };
+  }
+
+  private async flush() {
```

Comment:
After I upload an asset, I saw this method invoked. From the code, looks like it is only called from the instructor of `TypeSenseRepository`, can you explain this a bit?

Reply:
When you change an asset or an album it calls `searchRepository.index`, which adds the asset to an in-memory list. Every 5 seconds the list is sent to typesense as a batch upsert. Flush is called every 5 seconds (and the interval is setup in the constructor). Basically, this is batched updates (internal repository implementation) for performance reasons if there are lots of asset updates happening over a short period of time. If you run the CLI, it'll create lots of assets quickly, but this will only make http requests to typesense every 5 seconds with a list of assets.